### PR TITLE
Enhance cartservice's Redis reconnection policy

### LINKED ISSUE #561: Cartservice failing on skaffold run
Happening both locally and on the latest master branch deploy tests. 

```
+ cartservice-6df59b7fc-dg84z › server
cartservice-6df59b7fc-dg84z server Connecting to Redis: redis-cart:6379,ssl=false,allowAdmin=true,connectRetry=5
cartservice-6df59b7fc-dg84z server Unhandled exception. StackExchange.Redis.RedisConnectionException: It was not possible to connect to the redis server(s). UnableToConnect on redis-cart:6379/Interactive, Initializing/NotStarted, last: NONE, origin: BeginConnectAsync, outstanding: 0, last-read: 1s ago, last-write: 1s ago, keep-alive: 180s, state: Connecting, mgr: 10 of 10 available, last-heartbeat: never, global: 7s ago, v: 2.2.4.27433
cartservice-6df59b7fc-dg84z server    at StackExchange.Redis.ConnectionMultiplexer.ConnectImpl(ConfigurationOptions configuration, TextWriter log)
cartservice-6df59b7fc-dg84z server    at StackExchange.Redis.ConnectionMultiplexer.Connect(ConfigurationOptions configuration, TextWriter log)
cartservice-6df59b7fc-dg84z server    at cartservice.cartstore.RedisCartStore.EnsureRedisConnected()
cartservice-6df59b7fc-dg84z server    at cartservice.cartstore.RedisCartStore.InitializeAsync()
cartservice-6df59b7fc-dg84z server    at cartservice.Startup.ConfigureServices(IServiceCollection services)
cartservice-6df59b7fc-dg84z server    at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
cartservice-6df59b7fc-dg84z server    at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
cartservice-6df59b7fc-dg84z server    at Microsoft.AspNetCore.Hosting.MethodInfoExtensions.InvokeWithoutWrappingExceptions(MethodInfo methodInfo, Object obj, Object[] parameters)
cartservice-6df59b7fc-dg84z server    at Microsoft.AspNetCore.Hosting.ConfigureServicesBuilder.InvokeCore(Object instance, IServiceCollection services)
cartservice-6df59b7fc-dg84z server    at Microsoft.AspNetCore.Hosting.ConfigureServicesBuilder.<>c__DisplayClass9_0.<Invoke>g__Startup|0(IServiceCollection serviceCollection)
cartservice-6df59b7fc-dg84z server    at Microsoft.AspNetCore.Hosting.ConfigureServicesBuilder.Invoke(Object instance, IServiceCollection services)
cartservice-6df59b7fc-dg84z server    at Microsoft.AspNetCore.Hosting.ConfigureServicesBuilder.<>c__DisplayClass8_0.<Build>b__0(IServiceCollection services)
cartservice-6df59b7fc-dg84z server    at Microsoft.AspNetCore.Hosting.GenericWebHostBuilder.UseStartup(Type startupType, HostBuilderContext context, IServiceCollection services, Object instance)
cartservice-6df59b7fc-dg84z server    at Microsoft.AspNetCore.Hosting.GenericWebHostBuilder.<>c__DisplayClass13_0.<UseStartup>b__0(HostBuilderContext context, IServiceCollection services)
cartservice-6df59b7fc-dg84z server    at Microsoft.Extensions.Hosting.HostBuilder.CreateServiceProvider()
cartservice-6df59b7fc-dg84z server    at Microsoft.Extensions.Hosting.HostBuilder.Build()
cartservice-6df59b7fc-dg84z server    at <Program>$.<Main>$(String[] args)
- cartservice-6df59b7fc-dg84z › server
+ cartservice-6df59b7fc-dg84z › server
```

neither of the last two PRs modified cartservice or redis.

### PR DESCRIPTION
**Pull request by NimJay**
Fixes #561.

### Change Summary
* This pull-request enhances the way in which the `cartservice` Pod attempts to reconnect to the Redis Pod by:
  1. increasing the number of retries from 5 to 30
  1. waiting 1000+ milliseconds between each retry (instead of just 100+ milliseconds)

### Background

* The `cartservice` Pod talks to a `Redis` Pod.
* Occasionally, when we start the app with `skaffold dev` or `skaffold debug`, the `cartservice` Pod crashes with the following error (consequently, terminating `skaffold`):
```
[cartservice-7f67f7cd89-v4mrc server] Unhandled exception. System.ApplicationException: Wasn't able to connect to redis
```

### My Local Testing

Here's how I tested this pull-request on my local machine:

**1. Modify `skaffold.yaml`**

I altered the `skaffold.yaml` file such that it only builds/deploys `cartservice.yaml` and `redis.yaml`.
That is, I commented out all the other `image`s and `context`s, and changed
```yaml
deploy:
  kubectl:
    manifests:
    - ./kubernetes-manifests/**.yaml
```
to
```yaml
deploy:
  kubectl:
    manifests:
    - ./kubernetes-manifests/cartservice.yaml
    - ./kubernetes-manifests/redis.yaml
```

**2. Run skaffold**

I ran `skaffold debug --default-repo=gcr.io/nimjay-microservices-demo` many times with various configurations of the [Redis ReconnectRetryPolicy](https://stackexchange.github.io/StackExchange.Redis/Configuration#reconnectretrypolicy).

Number of Retries (`REDIS_RETRY_NUM`) | Minimum Milliseconds Between Retries ([`deltaBackOffMilliseconds`](https://github.com/StackExchange/StackExchange.Redis/blob/51d2946b7f1499605d90b95a00bcb8f776ff0b49/src/StackExchange.Redis/ExponentialRetry.cs#L19)) | # of times I ran `skaffold` | # of times `skaffold` erred | Notes
:---:|:---:|:---:|:---:|:---:
2 |  10 | 10 |  7 | -
5 | 100 | 10 | 2 | This is the existing configuration.
30 | 1000 | 10 | 0 | This is the configuration from this pull-request.

### How To Code-Review

If you want to be a bit rigorous with your code review, you can try what I describe above but only with:
1. 2 retries and 10ms gaps, until you see the error.
2. 30 retries with 1000ms gaps, a few times. It shouldn't err (at least not as often).

### CODE REVIEW

File: src/cartservice/src/cartstore/RedisCartStore.cs [hunk 1]
```diff
@@ -40,13 +40,13 @@ public RedisCartStore(string redisAddress)
             // Serialize empty cart into byte array.
             var cart = new Hipstershop.Cart();
             emptyCartBytes = cart.ToByteArray();
-            connectionString = $"{redisAddress},ssl=false,allowAdmin=true,connectRetry=5,abortConnect=false";
+            connectionString = $"{redisAddress},ssl=false,allowAdmin=true,abortConnect=false";
 
             redisConnectionOptions = ConfigurationOptions.Parse(connectionString);
 
-            // Try to reconnect if first retry failed (up to 5 times with exponential backoff)
```
Comment (NimJay) [inline src/cartservice/src/cartstore/RedisCartStore.cs:47]: I have reworded the comment so we don't have to update it every time `REDIS_RETRY_NUM` changes.

File: src/cartservice/src/cartstore/RedisCartStore.cs [hunk 2]
```diff
@@ -40,13 +40,13 @@ public RedisCartStore(string redisAddress)
             // Serialize empty cart into byte array.
             var cart = new Hipstershop.Cart();
             emptyCartBytes = cart.ToByteArray();
-            connectionString = $"{redisAddress},ssl=false,allowAdmin=true,connectRetry=5,abortConnect=false";
```
Comment (NimJay) [inline src/cartservice/src/cartstore/RedisCartStore.cs:43]: I remove `connectRetry=5` because it's redundant.

**Review by donmccasland — APPROVED** [review]

**Comment by bourgeoisor** [comment]
The column "Milliseconds Between Retries" is only accurate if we use a linear retry policy, but we are using an exponential retry policy (e.g. `new ExponentialRetry(1000)`), it may be misleading, though I do agree we should use an exponential policy. Perhaps renaming the column?

Do we have strong evidence that this PR permanently fixes the out-of-retries issue or just for "most" cases? I think in either case it can't hurt more and is worth bringing this change in.

**Review by bourgeoisor — APPROVED** [review]
LGTM!

**Comment by NimJay** [comment]
Thanks for the review, @donmccasland and @bourgeoisor.

**Regarding @bourgeoisor's comments:**

> The column "Milliseconds Between Retries" ... may be misleading

Column renamed to "Minimum Milliseconds Between Retries ([`deltaBackOffMilliseconds`](https://github.com/StackExchange/StackExchange.Redis/blob/51d2946b7f1499605d90b95a00bcb8f776ff0b49/src/StackExchange.Redis/ExponentialRetry.cs#L19))". :)

> Do we have strong evidence that this PR permanently fixes the out-of-retries issue or just for "most" cases? I think in either case it can't hurt more and is worth bringing this change in. 

Unfortunately, we don't have strong evidence that this is a permanent fix. And I agree, it's better to bring this fix in.

**Pull request merged**